test/server/config_validation/xds_verifier.cc

```diff
@@ -0,0 +1,325 @@
+#include "test/server/config_validation/xds_verifier.h"
+
+#include "common/common/logger.h"
+
+namespace Envoy {
+
+XdsVerifier::XdsVerifier(test::server::config_validation::Config::SotwOrDelta sotw_or_delta)
+    : num_warming_(0), num_active_(0), num_draining_(0), num_added_(0), num_modified_(0),
+      num_removed_(0) {
+  if (sotw_or_delta == test::server::config_validation::Config::SOTW) {
+    sotw_or_delta_ = SOTW;
+  } else {
+    sotw_or_delta_ = DELTA;
+  }
+  ENVOY_LOG_MISC(debug, "sotw_or_delta_ = {}", sotw_or_delta_);
+}
+
+/**
+ * get the route referenced by a listener
+ */
+std::string XdsVerifier::getRoute(const envoy::config::listener::v3::Listener& listener) {
+  envoy::config::listener::v3::Filter filter0 = listener.filter_chains()[0].filters()[0];
+  envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager conn_man;
+  filter0.typed_config().UnpackTo(&conn_man);
+  return conn_man.rds().route_config_name();
+}
+
+/**
+ * @return true iff the route listener refers to is in all_routes_
+ */
+bool XdsVerifier::hasRoute(const envoy::config::listener::v3::Listener& listener) {
+  return all_routes_.contains(getRoute(listener));
+}
+
+bool XdsVerifier::hasActiveRoute(const envoy::config::listener::v3::Listener& listener) {
+  return active_routes_.contains(getRoute(listener));
+}
+
+/**
+ * prints the currently stored listeners and their states
+ */
+void XdsVerifier::dumpState() {
+  ENVOY_LOG_MISC(debug, "Listener Dump:");
+  for (const auto& rep : listeners_) {
+    ENVOY_LOG_MISC(debug, "Name: {}, Route {}, State: {}", rep.listener.name(),
+                   getRoute(rep.listener), rep.state);
+  }
+}
+
+/*
+ * if you add a new listener that is not being updated, there are two cases:
+ * 1. the listener has not been added before
+ *    - if envoy knows about the listener's route, it will be added as active, else warming
+ * 1. a listener of the same name has been added before but it now draining
+ *    - if the routes match, the old listener will be added back to active
```

Comment:
This seems not quite right; a draining listener becomes active again? Or did you mean to say that the new listener becomes active right away given the route table is already known.

Reply:
Yep, you're right.  Should be new listener becomes active right away.